src/kakarot/library.cairo

```diff
@@ -355,23 +349,35 @@ namespace Kakarot {
         data: felt*,
     ) -> (return_data_len: felt, return_data: felt*) {
         alloc_locals;
-        let (success) = transfer(origin, to, value);
-        with_attr error_message("Kakarot: eth_call: failed to transfer {value} tokens to {to}") {
-            assert success = TRUE;
-        }
 
         if (to == 0) {
-            with_attr error_message("Kakarot: value should be 0 when deploying a contract") {
-                assert value = 0;
+            // TODO: read the nonce from the provided origin address, otherwise in view mode this will
+            // TODO: always use a 0 nonce
+            let (tx_info) = get_tx_info();
+            let (evm_contract_address) = CreateHelper.get_create_address(origin, tx_info.nonce);
+
+            let (success) = transfer(origin, evm_contract_address, value);
+            with_attr error_message(
+                    "Kakarot: eth_call: failed to transfer {value} tokens to {evm_contract_address}") {
+                assert success = TRUE;
             }
-            let (starknet_contract_address, evm_contract_address) = deploy_contract_account(
-                origin=origin, bytecode_len=data_len, bytecode=data
+
+            let (starknet_contract_address) = deploy_contract_account(
+                origin=origin,
+                evm_contract_address=evm_contract_address,
+                bytecode_len=data_len,
+                bytecode=data,
             );
             let (return_data) = alloc();
             assert [return_data] = evm_contract_address;
             assert [return_data + 1] = starknet_contract_address;
             return (2, return_data);
         } else {
+            let (success) = transfer(origin, to, value);
+            with_attr error_message(
+                    "Kakarot: eth_call: failed to transfer {value} tokens to {to}") {
+                assert success = TRUE;
```

Comment:
Not sure in Cairo0 but I think we can make `success` to be local and assert outside the if block to reduce code duplication

Reply:
it may create memory holes (to be verified though)

I think it's fine for now